{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Background and Motivation\n",
    "\n",
    "* The example here is chromosome 22 in the low copy repeat A region (LCRA)\n",
    "\n",
    "* This reference genome (hg38) looks as so:\n",
    "\n",
    "* REF\n",
    "\n",
    "* This region has 2 prominent features:\n",
    "\n",
    "1. It has 3 N-gaps (rather than ACTG it has \"N\"), indicating how nobody could assemble these portions of the region - it is unknown.\n",
    "\n",
    "2. The other feature is not obvious here. There are a lot of rearrangements and copy number variants. This makes most genomes become heterozyous and the 2 haplotypes are vastly different. Further, different genomes tend to be vastly different from one another.\n",
    "\n",
    "* Example:\n",
    "\n",
    "* 3 different genomes' LCRAs (vastly different ones)\n",
    "\n",
    "* For more info, see here: https://genome.cshlp.org/content/29/9/1389.full\n",
    "\n",
    "# Defining Critical Regions Through Ambiguity\n",
    "\n",
    "* Here, we generally define a critical region through ambiguous label patterns.\n",
    "\n",
    "* What this means is that if we mapped a molecule to a reigon with ambiguous label patterns, this same pattern may exist elsewhere in the reference (e.g., subtelomeres share modules between chromosomes and chromosome arms), and/or within the same haplotype via copy number increases, and/or within the same genome between its 2 haplotypes, where the 2 haplotypes differ by only a label or a few labels (mapping a molecule would not penalize it too much to not map to either and may pick the first one or molecule quality may affect the mapping).\n",
    "\n",
    "* Through experience, I know this red box is ambiguous\n",
    "\n",
    "* If a molecule maps to this red box, it maps to 2 loci in chr22.\n",
    "\n",
    "* It maps here:\n",
    "\n",
    "* LCRA\n",
    "\n",
    "* ...and here:\n",
    "\n",
    "* LCRD\n",
    "\n",
    "* How does this look in an XMAP?\n",
    "\n",
    "* XMAP\n",
    "\n",
    "* If I do not know the ambiguous regions beforehand, how do I define a critical region.\n",
    "\n",
    "* A couple ways:\n",
    "\n",
    "1. UCSC segmental duplication (SD) track as a BED file - can load this in Bionano Access and see where these are. The reason we choose SDs is because these are large and identical, so they are present even in our long molecules.\n",
    "    * This is a baseline for our critical region. Since optical maps are lower resolution than sequencing, there is no guarantee that that the DLE1 (or Nt.BspQI) labels are repetitive in a manner that would create multi-mapping or split-mapping. Thus, this is a baseline starting spot.\n",
    "\n",
    "2. Map molecules to a region (LCRA here) and parse the XMAP for loci where the molecule maps and/or split-maps.\n",
    "    * The example above, with the red box, shows how a molecule maps to LCRA AND LCRD with almost the same confidence score! Thus, we would define 18.74M - 18.88M as a critical region. The major advantage of this one vs number 1 is the region is much, much smaller.\n",
    "    \n",
    "* Now, if we choose number 2, we have to do the alignment check per genome. Why? Because what I did not show is that there is 1 other ambiguous SD module:\n",
    "\n",
    "* REF with gold and red.\n",
    "\n",
    "* The gold box indicates another module which has some labels also existing in LCRD.\n",
    "\n",
    "* Since rearrangements can leave this module out of some haplotypes and present in others, we would miss this ambiguous module if we only checked the critical region for one or a few genomes.\n",
    "\n",
    "* Thus, to account for variability between haplotypes (and genomes), if we choose the second method of defining a critical region, we have to map-check each genome.\n",
    "\n",
    "# Defining Critical Regions Through Unique Labels\n",
    "\n",
    "* Since an ambiguous region needs to be spanned, we have to anchor molecules in label patterns we know to be unique to a given region.\n",
    "\n",
    "* Thus, in addition to ambiguous label patterns, we choose regions flanking the ambiuity in the 5' and/or 3' end (where applicable - subtelomeres only have 1 anchor) where molecules map only to this label pattern and nowhere else and also do not split-map to other regions (SVs are an exception but ignore for now).\n",
    "\n",
    "* This step is simply the opposite to the one above; just ensure molecules are mapping to a label pattern flanking our ambiguity in a unique manner and the previous step defines this for us anyway - thus, we can skip this step.\n",
    "\n",
    "# Recruiting Molecules to the Newly-Defined Unique + Ambiguous Region\n",
    "\n",
    "* Now that we have defined the ambiguous region we must span, we can recruit molecules to the region to determine where they map and how they map (uniquely or ambiguously).\n",
    "\n",
    "* We have 3 ways:\n",
    "\n",
    "1. Molecules mapping to the reference\n",
    "\n",
    "2. Molecules mapping to contigs which map to our region\n",
    "    * Why? Because reference may be wrong or not indicative of a haplotype for this region and the de novo assembly may have captured some of the modules the reference missed.\n",
    "\n",
    "3. Non-mapping molecules\n",
    "\n",
    "# The Algorithm - First Checks\n",
    "\n",
    "* We can do 2 things to check before even running the algorithm\n",
    "\n",
    "1. de novo assembled contig's molecules\n",
    "    * Check molecules of each haplotype's contigs and if they are anchored and cross ambiguity and differ enough from one another, then we are done!\n",
    "    \n",
    "2. Check of there are completely spanning molecules\n",
    "    * If we have at least 2 completely spanning (e.g., 5' anchored, through ambiuity, 3' anchored) AND they differ from one another to discern 2 haplotypes (unless long stretch of homozyogus labels), then we are done!\n",
    "    * We can use these 2 molecules as the reference maps and map molecules to each and obtain our final molecule evidence!\n",
    "    \n",
    "* If neither applicable, then we can run the algorithm\n",
    "\n",
    "# The Algorithm - Running\n",
    "\n",
    "1. Split- and multi-map a genome's BNX file to our defined region\n",
    "    * scripts/complex_region_alignment.sh\n",
    "\n",
    "2. Separate aligned molecules into 4 groups (where applies): 5'-anchored, 3'-anchored, nested, and completely spanning\n",
    "    * scripts/separate_alignment.sh\n",
    "    * Here, we define an initial amount of anchor labels required to be anchored outside the ambiguous region\n",
    "    * We also define a set amount of complex labels to be in the ambiguous region\n",
    "\n",
    "3. Optional: Fix the nested molecules s/t the remaining molecules only map to this region and not elsewhere (ambiuous for haplotypes for this region)\n",
    "    * scripts/fix_nested.sh\n",
    "    * Note: for some regions this will remove all nested molecules, which is not what we want if they are required\n",
    "\n",
    "4. Extract query labels for anchor and complex regions for later comparisons\n",
    "    * scripts/extract_query_labels_anchor.sh\n",
    "    * We do this because we need to know the exact molecule labels corresponding to the anchor and complex labels\n",
    "\n",
    "5. Obtain anchor and complex label alignment matches for molecules (query labels from above guides this step)\n",
    "    * scripts/find_anchor_and_complex_alignment_matches.py\n",
    "    * We do this because we need to know the exact molecule labels corresponding to the anchor and complex labels\n",
    "    \n",
    "6. Filtering separated XMAPs of the above labels\n",
    "    * scripts/filter_maps_from_complex.sh\n",
    "    * We are simply removing any molecules mapped to the region, per separated group, if they do not meet the filtering criteria (X anchor labels and/or Y complex labels)\n",
    "    \n",
    "7. All-vs-all self-align molecules from all groups\n",
    "    * scripts/merge_into_one_cmap_self_align.sh\n",
    "    * As it states, we align each and every molecule to one another to ultimately see which molecules are similar to which other molecules and if we can eventually connect molecules from different groups\n",
    "    \n",
    "8. Filter self-mappings\n",
    "    * scripts/filter_molecules_self_alignment.py\n",
    "    * Here, we are ensuring mappings are correct\n",
    "    * We do not allow for molecules' anchor labels to map to other molecules' complex labels for instance\n",
    "    * We score mapped molecules based on complex labels only - each complex mapping score is summed to a final score (do NOT count anchor labels as this will artifically inflate same-separation molecules (e.g., one 5'-anchored vs another 5'-anchored molecule) to belong to one another and erroneously be categorized as the \"same\" haplotype)\n",
    "    * i.e., just because 2 molecules are 5'-anchored does not mean they are the same haplotype and counting anchor labels of 2 molecules with long anchors and shorter complex mappings (while still meeting the filtering criteria) may erroneuously map them to the same haplotype.\n",
    "\n",
    "9. Further filtering of self-mappings\n",
    "    * scripts/proportional_comparisons.py\n",
    "    * To increase specificity of molecules potentially belonging to the same haplotype, we obtain the distribution of the scores per molecule and obtain mean and std dev.\n",
    "    * We set molecules mapped to each molecule to 0 where mean - 1 std dev below the mean.\n",
    "    * This was done due to experimental evidence indicating false positive groupings of molecules into the same haplotype at the final clustering step.\n",
    "    * This reduces sensitivity but increases specificity\n",
    "\n",
    "10. Connect molecules mapped with scores above 0 between different groups\n",
    "    * sh scripts/long_haps.sh\n",
    "    * i.e., if a 5'-anchored molecule maps positively (score > 0) to a nested molecule, and this same nested molecule maps to a 3'-anchored molecule positively, then we place the 5'-, nested, and 3'-anchored molecules into the same group.\n",
    "\n",
    "11. Create matrix of all the scores and molecules for clustering\n",
    "    * sh scripts/all_vs_all_scores_matrix.sh\n",
    "\n",
    "12. K means (k=2 for each haplotype) the matrix of filtered aligment scores\n",
    "    * sh scripts/submit_kmeans.sh\n",
    "\n",
    "13. Generate final XMAPs - one per haplotype\n",
    "    * sh scripts/final_molecules.sh\n",
    "    * Can load each XMAP and the same QCMAP and RCMAP into Access to visuzlize\n",
    "    \n",
    "# TODO\n",
    "\n",
    "* Need an automated confirmation that the algorithm worked or failed\n",
    "\n",
    "* Need comparative method between genomes and ultimately group genomes based on haplotypes/modules\n",
    "\n",
    "* Better/improved methodology: what if we begin the algorithm, after the First Checks, with only the longest molecules and check if done, and if not, continue to next increment of molecule lengths?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
